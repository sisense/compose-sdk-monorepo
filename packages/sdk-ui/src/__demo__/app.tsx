import { Alert, Tab, Tabs } from '@mui/material';
import { Box } from '@mui/system';
import { useState } from 'react';
import { SisenseContextProvider } from '../components/SisenseContextProvider';
import { SisenseContextProviderProps } from '../props';
import { ChartsFromExampleApp } from './pages/ChartsFromExampleApp';
import { ECommerceDemo } from './pages/ECommerceDemo';
import { MiscDemo } from './pages/MiscDemo';
import { WidgetDemo } from './pages/WidgetDemo';

// This page is meant to enable faster iterations during development than
// using react-ts-demo or other demo apps that require a built sdk-ui
// Simply add a page to the `pages` array and it will show up in `yarn dev`
// Suggest adding a router or at least sessionStorage var for selectedTabIndex
// if this becomes popular
const pages = [WidgetDemo, ECommerceDemo, ChartsFromExampleApp, MiscDemo];

const { VITE_APP_SISENSE_URL, VITE_APP_SISENSE_USERNAME, VITE_APP_SISENSE_PASSWORD } = import.meta
  .env;

export function App() {
  const [selectedTabIndex, setSelectedTabIndex] = useState(0);

  const shouldShowAlert =
    !VITE_APP_SISENSE_URL || !VITE_APP_SISENSE_USERNAME || !VITE_APP_SISENSE_PASSWORD;

  const sisenseContextProps: SisenseContextProviderProps = {
    url: VITE_APP_SISENSE_URL ?? '',
    username: VITE_APP_SISENSE_USERNAME,
    password: VITE_APP_SISENSE_PASSWORD,
    defaultDataSource: 'Sample ECommerce',
  };

  return (
    <>
      {shouldShowAlert && (
        <Alert severity="warning">
          One or more expected env vars are missing. You may want to check your
          <code>.env.local</code> file.
        </Alert>
      )}
      <SisenseContextProvider {...sisenseContextProps}>
        <div style={{ width: '100vw', height: '100vh' }}>
          <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
            <Tabs
              value={selectedTabIndex}
              onChange={(e, value: number) => setSelectedTabIndex(value)}
            >
              {pages.map((page, i) => (
                <Tab key={i} label={page.name} sx={{ textTransform: 'none' }} />
              ))}
            </Tabs>
          </Box>
          {pages.map((Page, i) => (
            <div key={i} hidden={selectedTabIndex !== i}>
              {selectedTabIndex === i && (
                <Box sx={{ p: 3 }}>
                  <Page />
                </Box>
              )}
            </div>
          ))}
        </div>
      </SisenseContextProvider>
    </>
  );
}
