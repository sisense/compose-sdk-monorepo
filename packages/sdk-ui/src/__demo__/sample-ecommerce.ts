import { measureFactory } from '@sisense/sdk-data';

import { Commerce, Brand, Country, Category, DataSource } from './sample-ecommerce-autogenerated';

export { Commerce, Brand, Category, Country, DataSource };

/** adding predefined set of measures */
export const Measures = {
  SumCost: measureFactory.sum(Commerce.Cost).format('0,0$'),
  AvgCost: measureFactory.average(Commerce.Cost).format('0,0$'),
  SumRevenue: measureFactory.sum(Commerce.Revenue, 'Total Revenue').format('0,0$'),
  Quantity: measureFactory.sum(Commerce.Quantity, 'Total Quantity').format('0,0'),
  CountBrands: measureFactory.countDistinct(Commerce.BrandID).format('0,0'),
};
