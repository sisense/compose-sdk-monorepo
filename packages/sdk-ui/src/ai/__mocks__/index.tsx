import { SisenseContextProvider } from '@/sisense-context/sisense-context-provider';
import { mockToken, mockUrl } from '@/__mocks__/msw';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactNode } from 'react';
import AiContextProvider from '../ai-context-provider';

/**
 * A component that wraps an AI hook/component with all required providers.
 * For use in unit tests only.
 *
 * @internal
 */
export const AiTestWrapper = ({ children }: { children: ReactNode }) => (
  <SisenseContextProvider url={mockUrl} token={mockToken}>
    <AiContextProvider>
      <QueryClientProvider
        client={
          new QueryClient({
            defaultOptions: {
              queries: {
                retry: false,
              },
            },
          })
        }
      >
        {children}
      </QueryClientProvider>
    </AiContextProvider>
  </SisenseContextProvider>
);
