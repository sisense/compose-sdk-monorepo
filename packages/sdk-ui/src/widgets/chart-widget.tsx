/* eslint-disable @typescript-eslint/naming-convention */
import { ReactNode, useCallback, useMemo, useState, type FunctionComponent } from 'react';
import { Chart } from '../chart';
import { ChartProps, ChartWidgetProps, DrilldownWidgetProps, HighchartsOptions } from '../props';
import { asSisenseComponent } from '../decorators/component-decorators/as-sisense-component';
import { DynamicSizeContainer, getWidgetDefaultSize } from '../dynamic-size-container';
import { getDataSourceName } from '@sisense/sdk-data';
import { WidgetContainer } from './common/widget-container';
import { useSisenseContext } from '@/sisense-context/sisense-context';
import { DrilldownWidget } from './drilldown-widget';
import { useHighlightSelection } from './hooks/use-highlight-selection';
import { combineHandlers } from '../utils/combine-handlers';
import {
  createDrilldownToChartConnector,
  getDrilldownInitialDimension,
  isDrilldownApplicableToChart,
} from './common/drilldown-connector';
import { useSyncedDrilldownPaths } from './hooks/use-synced-hierarchies';
/**
 * The Chart Widget component extending the {@link Chart} component to support widget style options.
 * It can be used along with the {@link DrilldownWidget} component to support advanced data drilldown.
 *
 * @example
 * Example of using the `ChartWidget` component to
 * plot a bar chart of the `Sample ECommerce` data source hosted in a Sisense instance.
 * ```tsx
 * <ChartWidget
 *   dataSource={DM.DataSource}
 *   chartType="bar"
 *   dataOptions={{
 *     category: [DM.Category.Category],
 *     value: [measureFactory.sum(DM.Commerce.Revenue)],
 *     breakBy: [],
 *   }}
 * />
 * ```
 *
 * <img src="media://chart-widget-with-drilldown-example-1.png" width="800px" />
 * @param props - ChartWidget properties
 * @returns ChartWidget component representing a chart type as specified in `ChartWidgetProps.`{@link ChartWidgetProps.chartType | chartType}
 * @group Dashboarding
 */
export const ChartWidget: FunctionComponent<ChartWidgetProps> = asSisenseComponent({
  componentName: 'ChartWidget',
})((props) => {
  const { app } = useSisenseContext();
  const {
    chartType,
    dataSource = app?.defaultDataSource,
    dataOptions,
    styleOptions,
    drilldownOptions,
    highlightSelectionDisabled = false,
    highlights,
    description,
    onChange,
  } = props;
  const { width, height, ...styleOptionsWithoutSizing } = styleOptions || {};
  const defaultSize = getWidgetDefaultSize(chartType, {
    hasHeader: !styleOptions?.header?.hidden,
  });
  const [refreshCounter, setRefreshCounter] = useState(0);
  const isDrilldownApplicable = useMemo(
    () => isDrilldownApplicableToChart(chartType, dataOptions),
    [chartType, dataOptions],
  );
  const drilldownPaths = useSyncedDrilldownPaths({
    attribute: getDrilldownInitialDimension(chartType, dataOptions),
    dataSource: dataSource,
    drilldownPaths: drilldownOptions?.drilldownPaths,
    enabled: isDrilldownApplicable,
  });

  const isDrilldownEnabled = useMemo(() => {
    const hasDrilldownConfig =
      drilldownOptions?.drilldownSelections?.length ||
      drilldownOptions?.drilldownDimensions?.length ||
      drilldownPaths?.length;

    return hasDrilldownConfig && isDrilldownApplicable;
  }, [drilldownOptions, isDrilldownApplicable, drilldownPaths]);

  const highlightSelection = useHighlightSelection({
    chartType,
    dataOptions,
    enabled: !highlightSelectionDisabled && !isDrilldownEnabled && !highlights?.length,
  });

  const isAccessibilityEnabled = app?.settings.accessibilityConfig?.enabled;
  const applyWidgetDescriptionAsAccessibilityDescription: (
    options: HighchartsOptions,
  ) => HighchartsOptions = useCallback(
    (options) => {
      if (!isAccessibilityEnabled) {
        return options;
      }
      return {
        ...options,
        accessibility: {
          ...options.accessibility,
          description: description ?? options.accessibility?.description,
        },
      };
    },
    [description, isAccessibilityEnabled],
  );

  const onDrilldownWidgetChange = useCallback(
    ({ drilldownSelections }: Partial<DrilldownWidgetProps>) => {
      if (drilldownSelections) {
        onChange?.({
          drilldownOptions: {
            drilldownSelections,
          },
        });
      }
    },
    [onChange],
  );

  if (!chartType || !dataOptions) {
    return null;
  }

  const chartProps = {
    ...props,
    dataSet: dataSource,
    styleOptions: styleOptionsWithoutSizing,
    refreshCounter: refreshCounter,
    onDataPointClick: combineHandlers([
      highlightSelection.onDataPointClick,
      props.onDataPointClick,
    ]),
    onDataPointsSelected: combineHandlers([
      highlightSelection.onDataPointsSelected,
      props.onDataPointsSelected,
    ]),
    onBeforeRender: combineHandlers(
      [
        applyWidgetDescriptionAsAccessibilityDescription,
        highlightSelection.onBeforeRender,
        props.onBeforeRender,
      ],
      true,
    ),
  };

  const renderChart = (chartProps: ChartProps, topSlot?: ReactNode) => {
    return (
      <WidgetContainer
        {...props}
        topSlot={
          <>
            {props.topSlot}
            {topSlot}
          </>
        }
        dataSetName={dataSource && getDataSourceName(dataSource)}
        onRefresh={() => setRefreshCounter(refreshCounter + 1)}
      >
        <Chart {...chartProps} />
      </WidgetContainer>
    );
  };

  return (
    <DynamicSizeContainer
      defaultSize={defaultSize}
      size={{
        width,
        height,
      }}
    >
      {isDrilldownEnabled ? (
        <DrilldownWidget
          drilldownDimensions={drilldownOptions?.drilldownDimensions}
          drilldownPaths={drilldownPaths}
          initialDimension={getDrilldownInitialDimension(chartType, dataOptions)}
          drilldownSelections={drilldownOptions?.drilldownSelections || []}
          config={{
            isBreadcrumbsDetached: true,
          }}
          onChange={onDrilldownWidgetChange}
        >
          {(drilldownConnectProps) => {
            const { breadcrumbsComponent } = drilldownConnectProps;
            const withDrilldown = createDrilldownToChartConnector(drilldownConnectProps);

            return renderChart(withDrilldown(chartProps), breadcrumbsComponent);
          }}
        </DrilldownWidget>
      ) : (
        renderChart(chartProps)
      )}
    </DynamicSizeContainer>
  );
});
