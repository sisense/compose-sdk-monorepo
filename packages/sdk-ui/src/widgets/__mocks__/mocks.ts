import { QueryResultData } from '@sisense/sdk-data';

import * as DM from '../../__test-helpers__/sample-ecommerce';
import { mockPivotTableProps } from '../../pivot-table/__mocks__/mocks';
import { ChartWidgetProps, PivotTableWidgetProps } from '../../props';

export const mockPivotTableWidgetProps: PivotTableWidgetProps = {
  ...mockPivotTableProps,
  dataSource: DM.DataSource,
  title: 'Pivot table widget title',
  description: 'Pivot table widget desription',
  styleOptions: {
    width: 800,
    membersColor: true,
    headersColor: true,
    border: true,
    borderColor: 'lightgrey',
  },
} as PivotTableWidgetProps;

export const mockChartWidgetPropsForTable: ChartWidgetProps = {
  chartType: 'table',
  dataSource: DM.DataSource,
  title: 'Chart widget title',
  description: 'Chart widget desription',
  dataOptions: {
    columns: [DM.Commerce.AgeRange, DM.Commerce.Revenue],
  },
};

export const mockResolvedQuery: QueryResultData = {
  columns: [
    {
      name: 'Age Range',
      type: 'text',
    },
    {
      name: 'Revenue',
      type: 'number',
    },
  ],
  rows: [
    [
      {
        data: '19-24',
        text: '19-24',
        blur: false,
      },
      {
        data: 43.98411178588867,
        text: '43.9841117858887',
        blur: false,
      },
    ],
    [
      {
        data: '65+',
        text: '65+',
        blur: false,
      },
      {
        data: 62.803104400634766,
        text: '62.8031044006348',
        blur: false,
      },
    ],
  ],
};
