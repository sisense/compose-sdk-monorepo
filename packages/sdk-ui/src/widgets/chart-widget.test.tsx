/** @vitest-environment jsdom */
import { render } from '@testing-library/react';
import { measureFactory } from '@sisense/sdk-data';
import { type TFunction } from '@sisense/sdk-common';
import cloneDeep from 'lodash-es/cloneDeep';
import get from 'lodash-es/get';
import { mockChartWidgetPropsForTable, mockResolvedQuery } from './__mocks__/mocks';
import { useSisenseContextMock } from '../sisense-context/__mocks__/sisense-context';
import { ClientApplication } from '../app/client-application';
import { HttpClient, SsoAuthenticator } from '@sisense/sdk-rest-client';
import { ChartWidget } from './chart-widget';
import { executeQueryMock } from '@/query/__mocks__/execute-query';
import * as DM from '../__test-helpers__/sample-ecommerce';
import { MenuProvider } from '@/common/components/menu/menu-provider';
import { translation } from '@/translation/resources/en';

vi.mock('../query/execute-query');
vi.mock('../sisense-context/sisense-context');

const translateMock = vi.fn((key: string) => get(translation, key)) as unknown as TFunction;

vi.mock('react-i18next', async (importOriginal) => {
  const mod = await importOriginal<typeof import('react-i18next')>();
  return {
    ...mod,
    useTranslation: () => ({ t: translateMock }),
  };
});

describe('ChartWidget', () => {
  beforeEach(() => {
    const url = 'mock-url';
    useSisenseContextMock.mockReturnValue({
      app: {
        httpClient: new HttpClient(url, new SsoAuthenticator(url), 'test'),
        settings: {
          locale: 'en',
          dateConfig: {},
          trackingConfig: { enabled: false },
        },
      } as unknown as ClientApplication,
      tracking: { packageName: 'mock-package-name' },

      isInitialized: true,
    });
  });

  it('should render table widget', async () => {
    executeQueryMock.mockResolvedValue(mockResolvedQuery);
    const { container, findByLabelText } = render(
      <MenuProvider>
        <ChartWidget {...mockChartWidgetPropsForTable} />
      </MenuProvider>,
    );
    const table = await findByLabelText('table-root');
    expect(table).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('should render column widget with applied drilldown', async () => {
    // modifies data mock to match dataOptions with aggregated "Revenue" measure
    const dataMock = cloneDeep(mockResolvedQuery);
    dataMock.columns[1].name = '$measure0_sum Revenue';

    executeQueryMock.mockResolvedValue(dataMock);
    const { findByLabelText, findByText } = render(
      <MenuProvider>
        <ChartWidget
          {...{
            chartType: 'column',
            dataSource: DM.DataSource,
            title: 'Column Chart with Drilldown',
            dataOptions: {
              category: [DM.Commerce.Gender],
              value: [measureFactory.sum(DM.Commerce.Revenue)],
            },
            drilldownOptions: {
              drilldownDimensions: [],
              drilldownSelections: [
                {
                  points: [
                    {
                      categoryValue: 'Male',
                    },
                  ],
                  nextDimension: DM.Commerce.AgeRange,
                },
              ],
            },
          }}
        />
      </MenuProvider>,
    );

    // verifies rendered chart
    expect(await findByLabelText('chart-root')).toBeInTheDocument();

    // verifies existence of drilldown breadcrumbs
    expect(await findByText('Male')).toBeInTheDocument();
    expect(await findByText('Age Range (All)')).toBeInTheDocument();
  });
});
