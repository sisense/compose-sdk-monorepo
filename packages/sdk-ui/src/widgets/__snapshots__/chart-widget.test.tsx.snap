// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChartWidget > should render table widget 1`] = `
<div>
  <div
    style="width: 100%; height: 100%;"
  >
    <div
      style="width: 400px; height: 525px; position: relative;"
    >
      <div
        style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
      >
        <div
          class="csdk-w-full csdk-h-full csdk-overflow-hidden"
        >
          <div
            class="csdk-h-full"
            style="padding: 0px;"
          >
            <div
              class="csdk-h-full csdk-overflow-hidden"
              style="border-width: 0px; border-color: #9ea2ab; border-radius: 0; box-shadow: none; display: flex; flex-direction: column;"
            >
              <div
                class="csdk-flex csdk-flex-row csdk-items-center csdk-px-2 csdk-min-h-[32px]"
                style="background-color: rgb(255, 255, 255);"
              >
                <div
                  class="csdk-w-full csdk-whitespace-nowrap csdk-overflow-hidden"
                  style="text-align: left; color: rgb(91, 99, 114); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif; font-size: 15px;"
                >
                  Chart widget title
                </div>
                <div
                  class="csdk-ml-auto csdk-flex csdk-items-center"
                >
                  <button
                    aria-label="Click to view full details"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-av538e-MuiButtonBase-root-MuiIconButton-root"
                    data-mui-internal-clone-element="true"
                    style="color: rgb(0, 0, 0); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M11.5 20a7.5 7.5 0 1 1 0-15 7.5 7.5 0 0 1 0 15zm0-1a6.5 6.5 0 1 0 0-13 6.5 6.5 0 0 0 0 13zm-.5-8h1v5h-1v-5zm0-1v-.998h1V10h-1z"
                        fill="#5B6372"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
              <div
                style="flex-grow: 1; min-width: 0; min-height: 0; background-color: rgb(255, 255, 255);"
              >
                <div
                  style="width: 100%; height: 100%;"
                >
                  <div
                    style="width: 400px; height: 500px; position: relative;"
                  >
                    <div
                      style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
                    >
                      <div
                        aria-label="table-root"
                        style="display: flex; flex-direction: column; align-items: center; background-color: rgb(255, 255, 255);"
                      >
                        <div
                          class="_component_2a48df"
                        >
                          <div
                            class="_tableWrapper_4b88bf"
                            style="padding: 20px 20px;"
                          >
                            <div
                              aria-rowcount="3"
                              class="_table_4b88bf fixedDataTableLayout_main public_fixedDataTable_main"
                              role="grid"
                              style="height: 460px; width: 360px;"
                            >
                              <div
                                class="fixedDataTableLayout_rowsContainer"
                                style="height: 458px; width: 360px;"
                              >
                                <div
                                  class="fixedDataTableColumnResizerLineLayout_main fixedDataTableColumnResizerLineLayout_hiddenElem public_fixedDataTableColumnResizerLine_main"
                                  style="width: 0px; height: 460px; left: 0px;"
                                >
                                  <div
                                    class="fixedDataTableColumnResizerLineLayout_mouseArea"
                                    style="height: 460px;"
                                  />
                                </div>
                                <div
                                  class="fixedDataTableRowLayout_rowWrapper"
                                  style="width: 360px; height: 26px; z-index: 1; visibility: visible; transform: translate3d(0px,0px,0);"
                                >
                                  <div
                                    aria-rowindex="1"
                                    class="fixedDataTableRowLayout_main public_fixedDataTableRow_main fixedDataTableLayout_header public_fixedDataTable_header"
                                    role="row"
                                    style="width: 360px; height: 26px;"
                                  >
                                    <div
                                      class="fixedDataTableRowLayout_body"
                                    >
                                      <div
                                        class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                                        style="height: 26px; width: 0px; left: 0px;"
                                      >
                                        <div
                                          class="fixedDataTableCellGroupLayout_cellGroup"
                                          style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                                        />
                                      </div>
                                      <div
                                        class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                                        style="height: 26px; width: 360px; left: 0px;"
                                      >
                                        <div
                                          class="fixedDataTableCellGroupLayout_cellGroup"
                                          style="height: 26px; position: absolute; width: 252px; z-index: 0; transform: translate3d(0px,0px,0);"
                                        >
                                          <div
                                            class="fixedDataTableCellLayout_main public_fixedDataTableCell_main"
                                            role="columnheader"
                                            style="height: 26px; width: 126px; left: 0px;"
                                          >
                                            <div
                                              class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableHeader_4b88bf"
                                              style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                            >
                                              <div
                                                class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                              >
                                                <div
                                                  class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                                >
                                                  <div
                                                    class="public_fixedDataTableCell_cellContent"
                                                  >
                                                    <div
                                                      class="_headerCell_d9b667"
                                                    >
                                                      <div
                                                        class="_headerLabel_d9b667"
                                                      >
                                                        <div
                                                          class="_tableCellContent_4b88bf"
                                                          style="max-width: 90px;"
                                                        >
                                                          <div>
                                                            Age Range
                                                          </div>
                                                        </div>
                                                      </div>
                                                      <div
                                                        class="_sortIcon_d9b667 _caret_d9b667 _noSort_d9b667"
                                                      />
                                                    </div>
                                                  </div>
                                                </div>
                                              </div>
                                            </div>
                                          </div>
                                          <div
                                            class="fixedDataTableCellLayout_main fixedDataTableCellLayout_alignRight public_fixedDataTableCell_alignRight public_fixedDataTableCell_main"
                                            role="columnheader"
                                            style="height: 26px; width: 126px; left: 126px;"
                                          >
                                            <div
                                              class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableHeader_4b88bf"
                                              style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                            >
                                              <div
                                                class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                              >
                                                <div
                                                  class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                                >
                                                  <div
                                                    class="public_fixedDataTableCell_cellContent"
                                                  >
                                                    <div
                                                      class="_headerCell_d9b667"
                                                    >
                                                      <div
                                                        class="_headerLabel_d9b667"
                                                      >
                                                        <div
                                                          class="_tableCellContent_4b88bf"
                                                          style="max-width: 90px;"
                                                        >
                                                          <div>
                                                            Revenue
                                                          </div>
                                                        </div>
                                                      </div>
                                                      <div
                                                        class="_sortIcon_d9b667 _caret_d9b667 _noSort_d9b667"
                                                      />
                                                    </div>
                                                  </div>
                                                </div>
                                              </div>
                                            </div>
                                          </div>
                                        </div>
                                      </div>
                                      <div
                                        class=""
                                        style="left: 0px; height: 26px;"
                                      />
                                      <div
                                        class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                                        style="height: 26px; width: 0px; left: 360px;"
                                      >
                                        <div
                                          class="fixedDataTableCellGroupLayout_cellGroup"
                                          style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                                        />
                                      </div>
                                    </div>
                                  </div>
                                </div>
                                <div
                                  style="transform: translate3d(0px,26px,0);"
                                >
                                  <div
                                    class="fixedDataTableRowLayout_rowWrapper"
                                    style="width: 360px; height: 26px; z-index: 0; visibility: visible; transform: translate3d(0px,0px,0);"
                                  >
                                    <div
                                      aria-rowindex="2"
                                      class="fixedDataTableRowLayout_main public_fixedDataTableRow_main public_fixedDataTableRow_even public_fixedDataTable_bodyRow"
                                      role="row"
                                      style="width: 360px; height: 26px;"
                                    >
                                      <div
                                        class="fixedDataTableRowLayout_body"
                                      >
                                        <div
                                          class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                                          style="height: 26px; width: 0px; left: 0px;"
                                        >
                                          <div
                                            class="fixedDataTableCellGroupLayout_cellGroup"
                                            style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                                          />
                                        </div>
                                        <div
                                          class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                                          style="height: 26px; width: 360px; left: 0px;"
                                        >
                                          <div
                                            class="fixedDataTableCellGroupLayout_cellGroup"
                                            style="height: 26px; position: absolute; width: 252px; z-index: 0; transform: translate3d(0px,0px,0);"
                                          >
                                            <div
                                              class="fixedDataTableCellLayout_main public_fixedDataTableCell_main"
                                              role="gridcell"
                                              style="height: 26px; width: 126px; left: 0px;"
                                            >
                                              <div
                                                class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableCell_4b88bf"
                                                style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                              >
                                                <div
                                                  class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                                >
                                                  <div
                                                    class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                                  >
                                                    <div
                                                      class="public_fixedDataTableCell_cellContent"
                                                    >
                                                      <div
                                                        class="_tableCellContent_4b88bf"
                                                        style="max-width: 102px;"
                                                      >
                                                        <div>
                                                          19-24
                                                        </div>
                                                      </div>
                                                    </div>
                                                  </div>
                                                </div>
                                              </div>
                                            </div>
                                            <div
                                              class="fixedDataTableCellLayout_main fixedDataTableCellLayout_alignRight public_fixedDataTableCell_alignRight public_fixedDataTableCell_main"
                                              role="gridcell"
                                              style="height: 26px; width: 126px; left: 126px;"
                                            >
                                              <div
                                                class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableCell_4b88bf"
                                                style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                              >
                                                <div
                                                  class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                                >
                                                  <div
                                                    class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                                  >
                                                    <div
                                                      class="public_fixedDataTableCell_cellContent"
                                                    >
                                                      <div
                                                        class="_tableCellContent_4b88bf"
                                                        style="max-width: 102px;"
                                                      >
                                                        <div>
                                                          43.98
                                                        </div>
                                                      </div>
                                                    </div>
                                                  </div>
                                                </div>
                                              </div>
                                            </div>
                                          </div>
                                        </div>
                                        <div
                                          class=""
                                          style="left: 0px; height: 26px;"
                                        />
                                        <div
                                          class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                                          style="height: 26px; width: 0px; left: 360px;"
                                        >
                                          <div
                                            class="fixedDataTableCellGroupLayout_cellGroup"
                                            style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                                          />
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                  <div
                                    class="fixedDataTableRowLayout_rowWrapper"
                                    style="width: 360px; height: 26px; z-index: 0; visibility: visible; transform: translate3d(0px,26px,0);"
                                  >
                                    <div
                                      aria-rowindex="3"
                                      class="fixedDataTableRowLayout_main public_fixedDataTableRow_main public_fixedDataTableRow_highlighted public_fixedDataTableRow_odd public_fixedDataTable_bodyRow fixedDataTableLayout_hasBottomBorder public_fixedDataTable_hasBottomBorder"
                                      role="row"
                                      style="width: 360px; height: 26px;"
                                    >
                                      <div
                                        class="fixedDataTableRowLayout_body"
                                      >
                                        <div
                                          class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                                          style="height: 26px; width: 0px; left: 0px;"
                                        >
                                          <div
                                            class="fixedDataTableCellGroupLayout_cellGroup"
                                            style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                                          />
                                        </div>
                                        <div
                                          class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                                          style="height: 26px; width: 360px; left: 0px;"
                                        >
                                          <div
                                            class="fixedDataTableCellGroupLayout_cellGroup"
                                            style="height: 26px; position: absolute; width: 252px; z-index: 0; transform: translate3d(0px,0px,0);"
                                          >
                                            <div
                                              class="fixedDataTableCellLayout_main public_fixedDataTableCell_main"
                                              role="gridcell"
                                              style="height: 26px; width: 126px; left: 0px;"
                                            >
                                              <div
                                                class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableCell_4b88bf"
                                                style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                              >
                                                <div
                                                  class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                                >
                                                  <div
                                                    class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                                  >
                                                    <div
                                                      class="public_fixedDataTableCell_cellContent"
                                                    >
                                                      <div
                                                        class="_tableCellContent_4b88bf"
                                                        style="max-width: 102px;"
                                                      >
                                                        <div>
                                                          65+
                                                        </div>
                                                      </div>
                                                    </div>
                                                  </div>
                                                </div>
                                              </div>
                                            </div>
                                            <div
                                              class="fixedDataTableCellLayout_main fixedDataTableCellLayout_alignRight public_fixedDataTableCell_alignRight public_fixedDataTableCell_main"
                                              role="gridcell"
                                              style="height: 26px; width: 126px; left: 126px;"
                                            >
                                              <div
                                                class="fixedDataTableCellLayout_wrap1 public_fixedDataTableCell_wrap1 _tableCell_4b88bf"
                                                style="height: 26px; width: 126px; color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: "Open Sans","Roboto","Helvetica","Arial",sans-serif;"
                                              >
                                                <div
                                                  class="fixedDataTableCellLayout_wrap2 public_fixedDataTableCell_wrap2"
                                                >
                                                  <div
                                                    class="fixedDataTableCellLayout_wrap3 public_fixedDataTableCell_wrap3"
                                                  >
                                                    <div
                                                      class="public_fixedDataTableCell_cellContent"
                                                    >
                                                      <div
                                                        class="_tableCellContent_4b88bf"
                                                        style="max-width: 102px;"
                                                      >
                                                        <div>
                                                          62.8
                                                        </div>
                                                      </div>
                                                    </div>
                                                  </div>
                                                </div>
                                              </div>
                                            </div>
                                          </div>
                                        </div>
                                        <div
                                          class=""
                                          style="left: 0px; height: 26px;"
                                        />
                                        <div
                                          class="fixedDataTableCellGroupLayout_cellGroupWrapper"
                                          style="height: 26px; width: 0px; left: 360px;"
                                        >
                                          <div
                                            class="fixedDataTableCellGroupLayout_cellGroup"
                                            style="height: 26px; position: absolute; width: 0px; z-index: 2; transform: translate3d(0px,0px,0);"
                                          />
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
